futures-util/src/future/pinned.rs

```diff
@@ -0,0 +1,116 @@
+//! Definition of the `PinnedFut` adapter combinator
+
+// TODO: import `Pinned` and use it to make `Borrowed` immovable.
+use core::mem::PinMut;
+
+use futures_core::{Future, Poll};
+use futures_core::task;
+
+pub trait PinnedFnLt<'a, Data: 'a, Output> {
+    type Future: Future<Output = Output> + 'a;
+    fn apply(self, data: PinMut<'a, Data>) -> Self::Future;
+}
+
+pub trait PinnedFn<Data, Output>: for<'a> PinnedFnLt<'a, Data, Output> + 'static {}
+impl<Data, Output, T> PinnedFn<Data, Output> for T
+    where T: for<'a> PinnedFnLt<'a, Data, Output> + 'static {}
+
+impl<'a, Data, Output, Fut, T> PinnedFnLt<'a, Data, Output> for T
+where
+    Data: 'a,
+    T: FnOnce(PinMut<'a, Data>) -> Fut,
+    Fut: Future<Output = Output> + 'a,
+{
+    type Future = Fut;
+    fn apply(self, data: PinMut<'a, Data>) -> Self::Future {
+        (self)(data)
+    }
+}
+
+/// A future which borrows a value for an asynchronous lifetime.
+///
+/// Created by the `borrowed` function.
+#[must_use = "futures do nothing unless polled"]
+#[allow(missing_debug_implementations)]
+pub struct PinnedFut<Data: 'static, Output, F: PinnedFn<Data, Output>> {
+    fn_or_fut: FnOrFut<Data, Output, F>,
+    // TODO:
+    // marker: Pinned,
+    // Data, which may be borrowed by `fn_or_fut`, must be dropped last
+    data: Data,
```

Comment:
Maybe this should use `ManuallyDrop` so as to not rely on drop order?

(AFAIK drop order is actually "first field is dropped last", so this may even be wrong right now.)